### PR TITLE
techdebt: [CDS-77763]: Use service based streams for pipeline-service

### PR DESCRIPTION
**Pull request by sahilhindwani-harness**
## Describe your changes

The changes are to use service based streams for pipeline events. Currently we broadcast each and every of the event to each service and each service ignores that. This causes unncessary usage of resources in every service.

With this change, we are doing the following:
- Adding a consumer for every service  which listens on to the service specific topic
- A configuration to enable/disable this change.

 
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51972)

### CODE REVIEW

**Review by archit-harness — COMMENTED** [review]
> Reviewed 1 of 30 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51972)* status: 30 file(s) to review. 0 of 2 approvals obtained. 

___
*[`953-events-api/src/main/java/io/harness/eventsframework/EventsFrameworkConstants.java` line 90 at r1](https://reviewable.io/reviews/harness/harness-core/51972#-Nca547220NWIH7K_Zwk:-Nca547220NWIH7K_Zwl:bd777hn) ([raw file](https://github.com/harness/harness-core/blob/2fcbf1f6a079441633a2c7fbffb73daf3605d8df/953-events-api/src/main/java/io/harness/eventsframework/EventsFrameworkConstants.java#L90)):*
> ```Java
>   public static final String PIPELINE_NODE_START_EVENT_TOPIC = "pipeline_node_start";
> 
>   public static final String PIPELINE_NODE_START_EVENT_TOPIC_WITH_SERVICE_NAME = "%s_pipeline_node_start";
> ```

can we add %s at end, as its easy to read stream names from redis via prefix and not suffix
___
*[`clients/pipeline-service/pms-sdk/src/main/java/io/harness/pms/sdk/PmsSdkInitHelper.java` line 232 at r1](https://reviewable.io/reviews/harness/harness-core/51972#-Nca5Eiv8F35kxyvaiEB:-Nca5Eiv8F35kxyvaiEC:b-42pmgt) ([raw file](https://github.com/harness/harness-core/blob/2fcbf1f6a079441633a2c7fbffb73daf3605d8df/clients/pipeline-service/pms-sdk/src/main/java/io/harness/pms/sdk/PmsSdkInitHelper.java#L232)):*
> ```Java
>     switch (eventCategory) {
>       case INTERRUPT_EVENT:
>         if (streamPerServiceConfiguration.isMigrateInterrupt()) {
> ```

can we use only one config to migrate all streams simultaneously rather than per stream?
Plus i think we should change to take both streamNames from SDK.

and based on config/ff in pipeline-service -> we publish to both or just to old stream, that would be cleaner

**Comment by sahilhindwani-harness** [comment]
*[`clients/pipeline-service/pms-sdk/src/main/java/io/harness/pms/sdk/PmsSdkInitHelper.java` line 232 at r1](https://reviewable.io/reviews/harness/harness-core/51972#-Nca5Eiv8F35kxyvaiEB:-NcuGxuW2uNWBiXGifye:bbhtxbj) ([raw file](https://github.com/harness/harness-core/blob/2fcbf1f6a079441633a2c7fbffb73daf3605d8df/clients/pipeline-service/pms-sdk/src/main/java/io/harness/pms/sdk/PmsSdkInitHelper.java#L232)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

can we use only one config to migrate all streams simultaneously rather than per stream?
Plus i think we should change to take both streamNames from SDK.

and based on config/ff in pipeline-service -> we publish to both or just to old stream, that would be cleaner
</blockquote></details>

Migrating all streams at once could lead to sudden increase in stream length across modules. I have  added a config per stream so that the size does not increase all at once. We can test with one stream and later on move streams one by one.

**Comment by sahilhindwani-harness** [comment]
trigger uts

File: 120-ng-manager/config.yml [hunk 1]
```diff
@@ -551,3 +551,9 @@ webhookBranchHookEventHsqsDequeueConfig:
 
 useQueueServiceForWebhookTriggers: ${USE_QUEUE_SERVICE_FOR_WEBHOOK_TRIGGERS:-false}
 
+streamPerServiceConfiguration:
+  migrateNodeStart: ${PIE_MIGRATE_NODE_START:-false}
```
Comment (sahilhindwani-harness) [inline 120-ng-manager/config.yml:555]: Discussed with @hextechpal , will be making it a single config
cc- @archit-harness

**Comment by sahilhindwani-harness** [comment]
trigger review

**Comment by sahilhindwani-harness** [comment]
*[`953-events-api/src/main/java/io/harness/eventsframework/EventsFrameworkConstants.java` line 90 at r1](https://reviewable.io/reviews/harness/harness-core/51972#-Nca547220NWIH7K_Zwk:-NcwKXxODkZai5vf8lfN:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2fcbf1f6a079441633a2c7fbffb73daf3605d8df/953-events-api/src/main/java/io/harness/eventsframework/EventsFrameworkConstants.java#L90)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

can we add %s at end, as its easy to read stream names from redis via prefix and not suffix
</blockquote></details>

Done.

**Comment by sahilhindwani-harness** [comment]
trigger uts

**Comment by sahilhindwani-harness** [comment]
trigger uts

**Review by archit-harness — COMMENTED** [review]
> Reviewed 1 of 30 files at r1, 21 of 22 files at r3, 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51972)* status: 30 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. Waiting on team(s): sto.

**Comment by sahilhindwani-harness** [comment]
trigger uts

**Comment by sahilhindwani-harness** [comment]
trigger uts

**Review by archit-harness — DISMISSED** [review]
> Reviewed 3 of 4 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51972)* status: 30 file(s) to review. 1 of 2 approvals obtained. Waiting on team(s): sto.

**Comment by sahilhindwani-harness** [comment]
trigger uts

**Review by deepak-harness — APPROVED** [review]
> Reviewed 1 of 30 files at r1, 8 of 22 files at r3, 6 of 6 files at r4, 2 of 4 files at r5, 1 of 1 files at r6, 12 of 12 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51972)* status: 12 file(s) to review. 1 of 2 approvals obtained. Waiting on team(s): sto.

**Review by harshjain12 — APPROVED** [review]

**Comment by sahilhindwani-harness** [comment]
Some more changes started coming in this PR, therefore opened up a new PR: https://github.com/harness/harness-core/pull/53075

**Pull request closed without merging**